src/macos_native.rs

```diff
@@ -0,0 +1,892 @@
+            .store(true, Ordering::Relaxed);
+
+        {
+            let source = self.shared_state.source.lock().unwrap();
+
+            if let Some(source) = source.as_ref() {
+                unsafe { CFRunLoopSourceSignal(source.0.as_concrete_TypeRef()) }
+            }
+        }
+
+        {
+            let run_loop = self.shared_state.run_loop.lock().unwrap();
+
+            if let Some(run_loop) = run_loop.as_ref() {
+                unsafe {
+                    CFRunLoopWakeUp(run_loop.0.as_concrete_TypeRef());
+                }
+            }
+        }
+
+        // Wait for thread to shutdown
+        self.shared_state.shutdown_barrier.wait();
+
+        if let Some(handle) = self.reader_thread_handle.take() {
+            handle.join().unwrap();
+        }
+
+        if !self.shared_state.disconnected.load(Ordering::Relaxed) {
+            {
+                let device = self.shared_state.device.lock().unwrap();
+
+                // Error is also ignored in hid_close
+                let _result = device.close(self.open_options);
+            }
+        }
+
+        let mut input_reports = self.shared_state.input_reports.lock().unwrap();
+        input_reports.clear();
+    }
+}
+
+fn return_data(report: &[u8], buf: &mut [u8]) -> usize {
+    let copy_len = buf.len().min(report.len());
+
+    buf[..copy_len].copy_from_slice(&report[..copy_len]);
+
+    copy_len
+}
+
+impl HidDevice {
+    pub(crate) fn open(vid: u16, pid: u16, sn: Option<&str>) -> HidResult<Self> {
+        let devices = HidApiBackend::enumerate(Some(vid), Some(pid))?;
+
+        let target_sn = match sn {
+            Some(sn) => WcharString::String(sn.to_string()),
+            None => WcharString::None,
+        };
+
+        let target_dev = devices.iter().find(|dev| {
+            (dev.vendor_id == vid)
+                && (dev.product_id == pid)
+                && (sn.is_none() || (dev.serial_number == target_sn))
+        });
+
+        if let Some(dev) = target_dev {
+            Self::open_path(dev.path.as_c_str())
+        } else {
+            Err(HidError::HidApiError {
+                message: "device not found".into(),
+            })
+        }
+    }
+
+    pub(crate) fn open_path(device_path: &CStr) -> HidResult<Self> {
+        let entry =
+            open_service_registry_from_path(device_path).ok_or_else(|| HidError::HidApiError {
+                message: format!("Failed to open IOHIDDevice from path {device_path:?}"),
+            })?;
+
+        let device = IOHIDDevice::create(None, entry);
+
+        let ret = device.open(0);
+
+        if ret != kIOReturnSuccess {
+            return Err(HidError::HidApiError {
+                message: "failed to open IOHIDDevice from mach entry".into(),
+            });
+        }
+
+        let max_input_report_len = device
+            .get_i32_property("MaxInputReportSize")
+            .unwrap_or_default();
+
+        let run_loop_mode = format!("HIDAPI_{:p}", device.as_concrete_TypeRef());
```

Comment:
This isn't right, but [the `IOHIDDeviceUnscheduleFromRunLoop` `runLoopMode` parameter is unused anyway](https://github.com/apple-oss-distributions/IOKitUser/blob/b0b3f822b7507c265aa8a1e37c3100c03ca82039/hid.subproj/IOHIDDevice.c#L749-L753).

Reply:
Could you explain what you mean by not right? Using custom run modes seems to be possible according to the documentation at <https://developer.apple.com/documentation/corefoundation/cfrunloop?language=objc>.

Or do you mean that a custom run mode shouldn't be used?